### ISSUE TITLE
Detected 770 messages but deleted 0 message

### ISSUE THREAD

**Issue by TetraTheta** [victornpb/deleteDiscordMessages#24]
![주석 2019-11-22 003947](https://user-images.githubusercontent.com/37061801/69352640-9b7aec80-0cc0-11ea-9571-fdc0cb15a749.png)

I am trying to delete messages about 20 months ago(2018.02). When I press 'Start' button, result says there are 770 messages but deletes 0 message. I scrolled to that message and tried again. Same thing happened.

In this case, I didn't set 'Range' option. But When I set 'Range' option starting from very first message I've sent to last message I've sent, It stops to delete message at 26th deletion, so #23 might be related to.

```
Started at 2019. 11. 22. 오전 12:44:29
authorId="REDACTED" channelId="REDACTED" afterMessageId="REDACTED" beforeMessageId="REDACTED" hasLink=false hasFile=false
Total messages found: 704	(Messages in current page: 25, Author: 25, System: 0)	offset: 0
Delete delay: 100ms, Search delay: 100ms	Last Ping: 681ms, Average Ping: 681ms
Estimated time remaining: 0h 9m 12s
Waiting for your confirmation...
OK
0.14% (1/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.28% (2/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.43% (3/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.57% (4/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.71% (5/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.85% (6/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
0.99% (7/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.14% (8/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.28% (9/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.42% (10/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.56% (11/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.70% (12/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.85% (13/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.99% (14/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.13% (15/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.27% (16/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.41% (17/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.56% (18/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.70% (19/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.84% (20/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.98% (21/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
3.13% (22/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
3.27% (23/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
3.41% (24/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
3.55% (25/704)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 704	(Messages in current page: 0, Author: 0, System: 0)	offset: 0
Delete delay: 100ms, Search delay: 100ms	Last Ping: 301ms, Average Ping: 277ms
Estimated time remaining: 0h 4m 28s
Ended because API returned an empty page.
Ended at 2019. 11. 22. 오전 12:44:40! Total time: 0h 0m 10s
Delete delay: 100ms, Search delay: 100ms	Last Ping: 301ms, Average Ping: 277ms
Rate Limited: 0 times. Total time throttled: 0h 0m 0s.
Deleted 25 messages, 0 failed.
```

**Comment by Tehmaxx** [comment]
I get this exact issue as well. It finds all my messages but it just stops

Ended because API returned an empty page.

same as the above user

**Comment by LandyLane** [comment]
Having this issue as well.

**Comment by NoktyNx** [comment]
I also starting hitting this issue. Maybe the API was updated so we need a fix?

**Comment by nacorv** [comment]
it seems as though they are rate limiting by changing the search results, notably making the first page blank.

has anyone been able to figure out the endpoint called when scrolling through pages in the search history on the official client? i believe that there just needs to be another conditional inside of recurse() to handle the situation of no results but total > 0

**Comment by victornpb** [comment]
They announced in their changelog, that now you can search by oldest. They could also have changed something else in there. 

maybe this new sorting option could be also used to  delete messages more efficiently.

**Comment by victornpb** [comment]
Someone opened a pull request #25, but it doesn't look right, because it seems that there's no end condition now, and it will run forever, until you tell it to stop.

it seems that this line for some reason is triggering too soon.
https://github.com/victornpb/deleteDiscordMessages/blob/3835fe7864070c118e37a8fd7fcd1a5dc4b32e9c/deleteDiscordMessages.js#L273

**Comment by zyhp** [comment]
your 'offset' isn't increasing, so the 'page' is always the first one.

open fiddler and check the request when you change the search page, the 'offset' increases by 25 everytime.

all i did, cuz i needed a fast fix and didn't care about it being perfect or permanent, was to check if the page was empty like u do, and sum 25 to the offset, so it goes to the next page.

**Comment by zyhp** [comment]
to permanent fix it, just figure out what is the max offset you can go, and if the offset goes higher than it, return and close it, else, keep adding to the offset.

**Comment by redring600** [comment]
@zyhp I'm sorry, but I'm not as experienced as you.  Would you please explain what you did in simpler terms so I can get a temporary fix?  I'm running into this same error on a particular DM I need to deal with.

**Comment by nacorv** [comment]
look at the changes he made to the code. he only changed two lines

in the exit condition (api gives empty page) he changed it so it will not end(), but will instead add 25 to the offset and call recurse() again.

**Comment by redring600** [comment]
I feel like a complete idiot, and I'm sure I'm seen as that way, but I don't understand what lines were changed.  I am so sorry, I just have zero programming experience.  Would someone please be so kind as to give me something to paste and tell me where to paste it?

**Comment by nacorv** [comment]
https://github.com/victornpb/deleteDiscordMessages/blob/e2ee9c30867554d70bf3562370ea3ce4f571ffe8/deleteDiscordMessages.js
this is a link to zyhp's pull request.

paste his pull request instead of the current version.

it is IMPORTANT that you watch this thing work. when you get lots of empty results you MUST refresh the discord page to stop the script. hammering the api like that is very noticeable and very bad for your account

**Comment by zyhp** [comment]
i will see if i can push a solid fix later, as soon as i have time again i will take a look.

**Comment by redring600** [comment]
Ah, this is working @nacorv .  Thank you so much!  So very, very much!

**Comment by nacorv** [comment]
do not leave it unattended. stop it the moment you get a couple empty results in a row.

 it is important to realize the script can lead to account termination even when it is working correctly.

**Comment by redring600** [comment]
Gotcha.  It hit that point and I stopped it.  Should I wait X amount of time and start it again?

**Comment by nacorv** [comment]
if there are still messages in the dm then it is not done

the issue i am referring to is when you let it run after the last message is deleted. it is safe to run until then.

**Comment by redring600** [comment]
Oooh, that's odd.  It got a couple of page errors, so I stopped it, closed Chrome, re-opened it, and started up again.. and it was just empty page errors one after the other, so I stopped it again.  There are still many messages in the thread.

**Comment by redring600** [comment]
Gave it a few minutes and tried again and it's working.  Seriously, thank you all so much.

**Comment by victornpb** [comment]
> your 'offset' isn't increasing, so the 'page' is always the first one.
> 
> open fiddler and check the request when you change the search page, the 'offset' increases by 25 everytime.
> 
> all i did, cuz i needed a fast fix and didn't care about it being perfect or permanent, was to check if the page was empty like u do, and sum 25 to the offset, so it goes to the next page.

The way I wrote it, it was supposed to keep calling the first page, because as you delete messages, more and more messages gets shifted to the first page, so I only had to use the offset to skip messages that were system messages. It seems that the API no longer behaves like this.

I'm just too busy atm so I can't work on this rn, but I'm willing to accept pull requests once a more robust algorithm comes up.

**Comment by Deevits** [comment]
The last adjustment seems to make it work again. Good work!! Performance is pretty poor, but it does what it suppose to do: delete all messages :)

**Comment by cswl** [comment]
```
                if ( (offset * 25)  > grandTotal) {
                    log.warn('Ended because API returned an empty page.');
                    return end();

                }
                
               
                log.debug(`Cur offset ${offset} : total  : ${total}  ${grandTotal}`)
                // Forcefully increase offset owo
                offset = offset + 25;
     			return await recurse();
                 
            }
````
This one should work.. and stop hopefully

**Comment by victornpb** [comment]
I did some testing and it seems that discord is using 2 different search endpoints. I think I got it behaving just like the discord app. I ran a test on both DM channel and a Server (guild) channel.
The new endpoint for searching server channel messages requires you to pass a guildId, so I had to an extra field for that also. The good thing is that it only took me a few minutes.
And there's no need to increase the offset manually, as it would cause issues with system messages #15.

<details>
<summary>Test</summary>

```
Started at 12/1/2019, 5:11:30 PM
authorId="REDACTED" guildId="REDACTED" channelId="REDACTED" afterMessageId="REDACTED" beforeMessageId="REDACTED" hasLink=false hasFile=false
Total messages found: 132	(Messages in current page: 25, Author: 25, System: 0)	offset: 0
Delete delay: 100ms, Search delay: 100ms	Last Ping: 371ms, Average Ping: 371ms
Estimated time remaining: 0h 1m 2s
Waiting for your confirmation...
OK
0.76% (1/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
1.52% (2/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
2.27% (3/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
3.03% (4/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Being rate limited by the API for 39ms! Adjusted delete delay to 139ms.
Delete delay: 139ms, Search delay: 100ms	Last Ping: 247ms, Average Ping: 317ms
Cooling down for 78ms before retrying...
3.79% (5/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
4.55% (6/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
5.30% (7/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
6.06% (8/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Being rate limited by the API for 42ms! Adjusted delete delay to 181ms.
Delete delay: 181ms, Search delay: 100ms	Last Ping: 191ms, Average Ping: 272ms
Cooling down for 84ms before retrying...
6.82% (9/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
7.58% (10/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
8.33% (11/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
9.09% (12/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
9.85% (13/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
10.61% (14/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
11.36% (15/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
12.12% (16/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
12.88% (17/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
13.64% (18/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
14.39% (19/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
15.15% (20/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
15.91% (21/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
16.67% (22/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
17.42% (23/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
18.18% (24/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
18.94% (25/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
19.70% (26/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
20.45% (27/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 111	(Messages in current page: 21, Author: 21, System: 0)	offset: 0
Delete delay: 181ms, Search delay: 100ms	Last Ping: 322ms, Average Ping: 227ms
Estimated time remaining: 0h 0m 45s
21.21% (28/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
21.97% (29/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
22.73% (30/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
23.48% (31/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
24.24% (32/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
25.00% (33/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
25.76% (34/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
26.52% (35/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
27.27% (36/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
28.03% (37/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
28.79% (38/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
29.55% (39/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
30.30% (40/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
31.06% (41/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
31.82% (42/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
32.58% (43/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
33.33% (44/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
34.09% (45/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
34.85% (46/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
35.61% (47/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
36.36% (48/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 90	(Messages in current page: 21, Author: 21, System: 0)	offset: 0
Delete delay: 181ms, Search delay: 100ms	Last Ping: 341ms, Average Ping: 239ms
Estimated time remaining: 0h 0m 38s
37.12% (49/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
37.88% (50/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
38.64% (51/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
39.39% (52/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
40.15% (53/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
40.91% (54/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
41.67% (55/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
42.42% (56/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
43.18% (57/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
43.94% (58/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
44.70% (59/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
45.45% (60/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
46.21% (61/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
46.97% (62/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
47.73% (63/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
48.48% (64/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
49.24% (65/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
50.00% (66/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
50.76% (67/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
51.52% (68/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
52.27% (69/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 72	(Messages in current page: 18, Author: 18, System: 0)	offset: 0
Delete delay: 181ms, Search delay: 100ms	Last Ping: 348ms, Average Ping: 212ms
Estimated time remaining: 0h 0m 28s
53.03% (70/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
53.79% (71/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
54.55% (72/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
55.30% (73/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
56.06% (74/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
56.82% (75/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
57.58% (76/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
58.33% (77/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
59.09% (78/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
59.85% (79/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
60.61% (80/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
61.36% (81/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
62.12% (82/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
62.88% (83/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
63.64% (84/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
64.39% (85/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
65.15% (86/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
65.91% (87/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 47	(Messages in current page: 25, Author: 25, System: 0)	offset: 0
Delete delay: 181ms, Search delay: 100ms	Last Ping: 464ms, Average Ping: 309ms
Estimated time remaining: 0h 0m 23s
66.67% (88/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
67.42% (89/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
68.18% (90/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
68.94% (91/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
69.70% (92/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
70.45% (93/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
71.21% (94/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
71.97% (95/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
72.73% (96/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
73.48% (97/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
74.24% (98/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
75.00% (99/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
75.76% (100/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
76.52% (101/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
77.27% (102/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
78.03% (103/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
78.79% (104/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
79.55% (105/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
80.30% (106/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
81.06% (107/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
81.82% (108/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
82.58% (109/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
83.33% (110/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
84.09% (111/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
84.85% (112/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 26	(Messages in current page: 21, Author: 21, System: 0)	offset: 0
Delete delay: 181ms, Search delay: 100ms	Last Ping: 414ms, Average Ping: 290ms
Estimated time remaining: 0h 0m 12s
85.61% (113/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
86.36% (114/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
87.12% (115/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
87.88% (116/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
88.64% (117/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
89.39% (118/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
90.15% (119/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
90.91% (120/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
91.67% (121/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
92.42% (122/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
93.18% (123/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
93.94% (124/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
94.70% (125/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
95.45% (126/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
96.21% (127/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
96.97% (128/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
97.73% (129/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
98.48% (130/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
99.24% (131/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
100.00% (132/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
100.76% (133/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 5	(Messages in current page: 1, Author: 1, System: 0)	offset: 0
Delete delay: 181ms, Search delay: 100ms	Last Ping: 278ms, Average Ping: 273ms
Estimated time remaining: 0h 0m 2s
101.52% (134/132)	Deleting ID:REDACTED REDACTED (REDACTED): REDACTED	
Searching next messages in 100ms...	
Total messages found: 0	(Messages in current page: 0, Author: 0, System: 0)	offset: 0
Delete delay: 181ms, Search delay: 100ms	Last Ping: 256ms, Average Ping: 266ms
Estimated time remaining: 0h 0m 0s
Ended at 12/1/2019, 5:12:48 PM! Total time: 0h 1m 17s
Delete delay: 181ms, Search delay: 100ms	Last Ping: 256ms, Average Ping: 266ms
Rate Limited: 2 times. Total time throttled: 0h 0m 0s.
Deleted 134 messages, 0 failed.
```
</details>

**Comment by nacorv** [comment]
incredible. thanks for the fast fix

**Comment by victornpb** [comment]
@nacorv thank you for all the feedback.
I'll let more people manifest to see if the last release fixed for them before closing this issue. 
Testing is sorta hard, because I have to first generate hundreds of messages, and sometimes it works on my end.

I noticed the search results sometimes are not "realtime" and have some kind of caching, which sometimes causes issues, because as we are deleting messages, we are also changing the search results.

I don't know exactly what kind of architecture discord uses to index messages, but it is very likely that this is a computationally expensive process, so they probably have some sort of speed layer combined with a long term layer. To circumvent this I could change the algorithm to first discover all messages and then start deleting them, but there are also trade offs in doing that.

**Comment by nacorv** [comment]
they use elasticsearch on the backend for this. I'm not very familiar with how that works but i believe that mass deletions do cause changes to the historical index.

additionally, bot accounts do deletion differently. they are only limited to 2 weeks of history and can either delete individual messages, ranges of them, or the bulk deletion endpoint which throws them away in groups of 100(?). i think the conservative limits of bot accounts points toward the resource intensiveness of mass search and deletion on discord's end.

**Comment by jumphil** [comment]
New release fully fixed it for me, I just deleted thousands of messages on different servers. Thanks for your work.
